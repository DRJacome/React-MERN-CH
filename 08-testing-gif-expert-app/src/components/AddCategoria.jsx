import React, { useState } from 'react'
import PropTypes from "prop-types";

/* Como parámetro se envía desestructurada la propiedad "{setCategorias}" para 
 * establecer comunicación entre el padre "GifExpertApp" y el hijo "AddCategoria"
*/
export const AddCategoria = ({ setCategorias }) => {
  const [inputValue, setInputValue] = useState('')

  const handleInputChange = (e) => {
    setInputValue(e.target.value)
  }

  const handleSubmit = (e) => {
    e.preventDefault()

    if (inputValue.trim().length > 2) {
      /* setCategorias(cat => [inputValue, ...cat]) */
      setCategorias(cat => [inputValue, ...cat])
      setInputValue('')
    }

    //console.log('Submit hecho');
  }
  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input
        type='text'
        value={inputValue}
        onChange={handleInputChange}
      />
    </form>
  )
}

/* PropTypes se usa para establecer reglas de requerimientos para la aplicación.
 * En este caso se usa "setCategorias: PropTypes.func.isRequired" para marcar una
 * regla que requiera que la función "setCaterorias" debe ser enviada desde 
 * el padre "GifExpertApp".
 */
AddCategoria.propTypes = {
  setCategorias: PropTypes.func.isRequired
  /* setCategorias: PropTypes.func */
}