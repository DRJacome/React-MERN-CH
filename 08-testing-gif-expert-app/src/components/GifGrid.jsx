import React from 'react'
import PropTypes from 'prop-types'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from "./GifGridItem"
//import { getGifs } from "../helpers/getGifs"

export const GifGrid = ({ categoria }) => {
  const { data: imagenes, loading } = useFetchGifs(categoria)

  /*   useEffect(() => {
      getGifs(categoria)
        .then(
        )
      /*  El segundo argumento del useEffect es llamado "Lista/array de dependencias []" y 
      * y objetivo es controlar cuándo se renderiza el componente. Si el array está vacío,
      * el componente sólo se renderizará por primera vez, evitando que se renderice continuamente.
      
    }, [categoria])
   */

  return (
    <>
      <h3 className='animate__animated animate__fadeIn'>{categoria}</h3>
      {loading && <p className='carta animate__animated animate__flash'>Cargando</p>}
      {<div className="contenedorCartas">
        {imagenes.map(img => (
          <GifGridItem
            key={img.id}
            {...img}
          />
        ))}
      </div>}
    </>
  )
}

GifGrid.propTypes = {
  categoria: PropTypes.string.isRequired
}