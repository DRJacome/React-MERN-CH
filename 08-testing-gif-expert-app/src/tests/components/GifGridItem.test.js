import React from 'react'
import { shallow } from 'enzyme'
import { GifGridItem } from '../../components/GifGridItem'

describe('Pruebas sobre <GifGridItem/>', () => {

  const title = 'Un título'
  const url = 'https://localhost/algo.jpg'

  /* Test renderizando el componente y sus propiedades. */
  const wrapper = shallow(<GifGridItem title={title} url={url} />)

  test('Debe mostrar el componente correctamente', () => {
    /* Test renderizando solo el componente. */
    const wrapper = shallow(<GifGridItem />)

    expect(wrapper).toMatchSnapshot()
  })

  test('Debe recibir las propiedades "title" y "url"', () => {

    expect(wrapper).toMatchSnapshot();
  })

  test('Debe obtener la imagen y el alt de los props', () => {
    const img = wrapper.find('img')
    console.log(img.prop('src'));

    expect(img.prop('src')).toBe(url);
    expect(img.prop('alt')).toBe(title);
  })

  test('Debe tener un párrafo con la propiedad "title"', () => {
    const p = wrapper.find('p')
    console.log(p.text());

    expect(p.text().trim()).toBe(title);
  })

  test('Debe tener la clase CSS animate__fadeIn', () => {
    const div = wrapper.find('div')
    const className = div.prop('className')
    console.log(div.prop('className'))


    expect(className.includes('animate__fadeIn')).toBe(true);
  })
})