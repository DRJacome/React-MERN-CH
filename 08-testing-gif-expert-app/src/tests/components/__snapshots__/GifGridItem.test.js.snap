// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <GifGridItem/> Debe mostrar el componente correctamente 1`] = `
<div
  className="carta animate__animated animate__fadeIn"
>
  <img />
  <p />
</div>
`;

exports[`Pruebas sobre <GifGridItem/> Debe recibir las propiedades "title" y "url" 1`] = `
<div
  className="carta animate__animated animate__fadeIn"
>
  <img
    alt="Un título"
    src="https://localhost/algo.jpg"
  />
  <p>
    Un título
  </p>
</div>
`;
